### PR TITLE
Add storybook hook example

### PR DESCRIPTION
**Pull request by mocca102**

### CODE REVIEW

File: src/hooks/usePropsGetters/useOpenTextInputProps.ts
Comment (esezen) [inline src/hooks/usePropsGetters/useOpenTextInputProps.ts:31]: React Question: Would changing this to `currentQuestionData.type` help here?

File: src/hooks/usePropsGetters/useOpenTextInputProps.ts
Comment (esezen) [inline src/hooks/usePropsGetters/useOpenTextInputProps.ts:54]: Why are we adding inline styles here?

File: src/hooks/usePropsGetters/useOpenTextInputProps.ts
Comment (esezen) [inline src/hooks/usePropsGetters/useOpenTextInputProps.ts:58]: Same as above

File: src/hooks/useQuiz.ts
Comment (esezen) [inline src/hooks/useQuiz.ts:27]: Is this intentional?

**Review by esezen — COMMENTED** [review]
What a change 🚀 
I only reviewed the implementation and not the stories but this is looking great so far. I left some comments here and there, most of them are clarification questions rather than change requests. 


---------------

Since this is a big change, I think it would be great if we can go over this during a call. Would you be able to schedule something among you, me, @deomsj + whoever else is interested?

**Review by esezen — COMMENTED** [review]
It looks like something happened to the "Continue Quiz" modal 🤔 
image

File: src/components/CioQuiz/quizLocalReducer.ts
Comment (deomsj) [inline src/components/CioQuiz/quizLocalReducer.ts:96]: `answerInputs: { ...newAnswerInputs },`
=>
`answerInputs: newAnswerInputs,`

File: src/components/RedoButton/RedoButton.tsx
Comment (deomsj) [inline src/components/RedoButton/RedoButton.tsx:15]: instead of disabling, I'd prefer that we add an attribute of `type="button"`

File: src/components/ResultCtaButton/ResultCtaButton.tsx
Comment (deomsj) [inline src/components/ResultCtaButton/ResultCtaButton.tsx:17]: instead of disabling, I'd prefer that we add an attribute of type="button"

File: src/constants.ts
Comment (deomsj) [inline src/constants.ts:37]: useCioAutocomplete 👀

File: src/components/ResultCtaButton/ResultCtaButton.tsx
Comment (deomsj) [reply src/components/ResultCtaButton/ResultCtaButton.tsx:17]: never mind, I see that we are returning this attr from our `getAddToCartButtonProps` function

File: src/hooks/usePropsGetters/useOpenTextInputProps.ts
Comment (deomsj) [reply src/hooks/usePropsGetters/useOpenTextInputProps.ts:31]: yup, changing this to `currentQuestionData.type` would cause this to effect to run only when the value of  `currentQuestionData.type`  changes as opposed to every time that `currentQuestionData` references a new object

File: src/hooks/usePropsGetters/usePreviousQuestionButtonProps.ts
Comment (deomsj) [inline src/hooks/usePropsGetters/usePreviousQuestionButtonProps.ts:18]: instead of limiting options for consumer of this lib, I'd prefer that we go with something like this:
```js
className={`cio-question-back-button ${buttonDisabled ? 'disabled' : ''}`}
```

File: src/hooks/useQuizEvents/useQuizAnswerChangeHandler.ts
Comment (deomsj) [inline src/hooks/useQuizEvents/useQuizAnswerChangeHandler.ts:41]: can we simplify this:
```js
            type:
              currentQuestion!.next_question.type === QuestionTypes.SingleSelect
                ? QuestionTypes.SingleSelect
                : QuestionTypes.MultipleSelect,
```
to this:
```js
  type: currentQuestion!.next_question.type,
```

File: src/hooks/useQuizEvents/useQuizNextClick.ts
Comment (deomsj) [inline src/hooks/useQuizEvents/useQuizNextClick.ts:18]: `currentAnswerInput?.value && currentAnswerInput?.value?.length`
=>
`currentAnswerInput?.value?.length`

File: src/hooks/usePropsGetters/useSelectInputProps.ts
Comment (deomsj) [inline src/hooks/usePropsGetters/useSelectInputProps.ts:71]: similar to comment above: https://www.30secondsofcode.org/react/s/use-effect-primitive-dependencies/

File: src/hooks/useQuizState/useQuizLocalState.ts
Comment (deomsj) [inline src/hooks/useQuizState/useQuizLocalState.ts:25]: React guarantees that dispatch function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.

[source](https://legacy.reactjs.org/docs/hooks-reference.html)

File: src/utils.tsx
Comment (deomsj) [inline src/utils.tsx:184]: let's tell eslint that we are doin this on purpose, so it knows not to warn us about it :)
Screenshot 2023-06-26 at 1 44 16 PM

File: src/components/CioQuiz/actions.ts
Comment (mocca102) [reply src/components/CioQuiz/actions.ts:1]: It wouldn't cause a problem with types but it's just ugly. To make it go away we would have to move a lot of types around which I felt it would get uglier so I just silenced it. But I can give it another stab.

File: src/hooks/useQuiz.ts
Comment (mocca102) [reply src/hooks/useQuiz.ts:27]: Will remove it before merging. But it's good for debugging now

File: src/hooks/usePropsGetters/useOpenTextInputProps.ts
Comment (mocca102) [reply src/hooks/usePropsGetters/useOpenTextInputProps.ts:31]: It doesn't really matter much here since when type change is when currentQuestionData change they happen together

File: src/stories/Quiz/Hooks/index.tsx
Comment (deomsj) [inline src/stories/Quiz/Hooks/index.tsx:104]: let's avoid inline styles

File: src/stories/Quiz/Hooks/index.tsx
Comment (deomsj) [inline src/stories/Quiz/Hooks/index.tsx:124]: lets avoid inline styles

File: src/stories/Quiz/Hooks/index.tsx
Comment (deomsj) [inline src/stories/Quiz/Hooks/index.tsx:282]: style

File: src/hooks/usePropsGetters/useSelectInputProps.ts
Comment (mocca102) [reply src/hooks/usePropsGetters/useSelectInputProps.ts:71]: Wait I remembered. There can be two questions after each other with the same question type. Say two cover questions. This way the effect won't run even though the currentQuestionData changed.

File: src/components/CioQuiz/actions.ts
Comment (esezen) [reply src/components/CioQuiz/actions.ts:1]: I'm assuming you gave it another go and it's still ugly. @deomsj any concern with this?

File: src/components/CioQuiz/quizLocalReducer.ts
Comment (esezen) [reply src/components/CioQuiz/quizLocalReducer.ts:1]: @mocca102  I see a 👍 , does that mean yes it's because of that but no there isn't anything else we can use?

File: src/hooks/usePropsGetters/useOpenTextInputProps.ts
Comment (esezen) [reply src/hooks/usePropsGetters/useOpenTextInputProps.ts:31]: But wouldn't that allow us to remove `react-hooks/exhaustive-deps`?

File: src/hooks/usePropsGetters/useOpenTextInputProps.ts
Comment (esezen) [reply src/hooks/usePropsGetters/useOpenTextInputProps.ts:54]: @mocca102

File: src/hooks/usePropsGetters/useOpenTextInputProps.ts
Comment (esezen) [reply src/hooks/usePropsGetters/useOpenTextInputProps.ts:58]: Again, same as above

File: src/components/CioQuiz/actions.ts
Comment (deomsj) [reply src/components/CioQuiz/actions.ts:1]: tech debt worth cleaning fo sho. 

I expect that this could only cause issues at build time, not during run time. so.... as long as we are happy with our test coverage, then this issue can live another day and be tackled as a separate tech debt ticket focused on resolving circular-imports

File: src/hooks/usePropsGetters/useOpenTextInputProps.ts
Comment (deomsj) [reply src/hooks/usePropsGetters/useOpenTextInputProps.ts:31]: yup and the underlying reasoning for WHY this'll make the linter happy is that the linter is enforcing the pattern that will help to avoid difficult to diagnose performance issue that can result from using objects in dependency arrays. 

We should only use primitives in dependency arrays

File: src/hooks/usePropsGetters/usePreviousQuestionButtonProps.ts
Comment (esezen) [reply src/hooks/usePropsGetters/usePreviousQuestionButtonProps.ts:18]: @mocca102 thoughts on the above?

File: src/hooks/usePropsGetters/useSelectInputProps.ts
Comment (esezen) [reply src/hooks/usePropsGetters/useSelectInputProps.ts:71]: Can we depend on id or something primitive that changes?

File: src/stories/Quiz/Hooks/index.tsx
Comment (esezen) [reply src/stories/Quiz/Hooks/index.tsx:104]: @mocca102

**Review by esezen — COMMENTED** [review]
Left some comments in the existing comment threads

File: src/components/CioQuiz/quizLocalReducer.ts
Comment (mocca102) [reply src/components/CioQuiz/quizLocalReducer.ts:1]: I changed the syntax [here](https://github.com/Constructor-io/constructorio-ui-quizzes/pull/76/commits/97cd84fb50adbc84640cb60a8d2964227d674b62#diff-85839113a3c54921f0066214f23b6a56a79617a4f3743176b21eb6ac598800f0R63) apologies forgot to remove the comment

File: src/hooks/usePropsGetters/useSelectInputProps.ts
Comment (esezen) [inline src/hooks/usePropsGetters/useSelectInputProps.ts:62]: Same as above

File: src/hooks/usePropsGetters/useSelectInputProps.ts
Comment (esezen) [inline src/hooks/usePropsGetters/useSelectInputProps.ts:89]: Can we depend on something primitive here?

File: src/stories/Quiz/Hooks/index.tsx
Comment (esezen) [inline src/stories/Quiz/Hooks/index.tsx:10]: This shouldn't look like this, right?

image

**Review by esezen — CHANGES_REQUESTED** [review]
Thanks for the changes, requested some more 😈 

This is shaping up nicely, let me know what do you think about the comments. After that we can get another review from @deomsj

**Review by esezen — APPROVED** [review]
This is looking good. 
Let's get another review from @deomsj

**Review by deomsj — CHANGES_REQUESTED** [review]
Wow. This has come such a long way. Functionality and code changes in this PR are looking great, @mocca102!!!

I flagged a few quick requested changes as inline comments. Please let me know when/what I can clarify / support/ re-review

**Review by deomsj — APPROVED** [review]
🔥 Great work on this @mocca102!!!

This is working great on my end and code looks ready to merge!

File: src/stories/Quiz/Component/QuestionTypes/MultipleSelectQuestion.stories.tsx
Comment (mocca102) [inline src/stories/Quiz/Component/QuestionTypes/MultipleSelectQuestion.stories.tsx:59]: Removing interaction tests from single components and only testing presentation

File: src/stories/Quiz/tests/index.stories.tsx
Comment (mocca102) [inline src/stories/Quiz/tests/index.stories.tsx:130]: Single select question doesn't need a continue click

File: src/hooks/usePropsGetters/useSelectInputProps.ts
Comment (esezen) [inline src/hooks/usePropsGetters/useSelectInputProps.ts:24]: Just out of curiosity, what's the benefit of `useRef` instead of `useState` here?

File: src/hooks/usePropsGetters/useSelectInputProps.ts
Comment (esezen) [inline src/hooks/usePropsGetters/useSelectInputProps.ts:80]: Can you add a comment here explaining what this effect does and why `answerInputs` is in the dependency array?

**Review by esezen — COMMENTED** [review]
@mocca102 this is looking good! I'll merge it shortly

File: src/hooks/usePropsGetters/useSelectInputProps.ts
Comment (mocca102) [reply src/hooks/usePropsGetters/useSelectInputProps.ts:24]: Will leave it as useRef instead of useState since we don't need this to cause rerenders and only we want to keep track of whether singleSelectClicked or not

**Pull request merged**
